federation/pkg/federation-controller/util/federated_informer.go

```diff
@@ -146,9 +164,16 @@ func NewFederatedInformer(federationClient federation_release_1_4.Interface, tar
 					return
 				}
 				if isClusterReady(oldCluster) != isClusterReady(curCluster) || !reflect.DeepEqual(oldCluster.Spec, curCluster.Spec) {
+
+					if clusterLifecycle.ClusterUnavailable != nil {
+						clusterLifecycle.ClusterUnavailable(oldCluster, federatedInformer, federatedInformer.GetTargetStore())
+					}
 					federatedInformer.deleteCluster(oldCluster)
 					if isClusterReady(curCluster) {
 						federatedInformer.addCluster(curCluster)
+						if clusterLifecycle.ClusterAvailable != nil {
+							clusterLifecycle.ClusterAvailable(curCluster, federatedInformer, federatedInformer.GetTargetStore())
+						}
```

Comment:
Let me ask one thing - what happens if you're changing cluster spec? Can this be done without available->not available -> available transition? If so, sending those events to users may be super misleading, because they will do some actions unnecessary.